lib/client/indexer/dao.ts

```diff
@@ -1,12 +1,166 @@
+  }
   async fetchAllDaos(): Promise<DaoAggregate[]> {
-    return [];
+    try {
+      const { dao_details } = await this.graphqlClient.query?.({
+        dao_details: {
+          // pass arguments to the query
+          name: true,
+          slug: true,
+          url: true,
+          github: true,
+          x_account: true,
+          description: true,
+          image_url: true,
+          daos: {
+            program_acct: true,
+            dao_acct: true,
+            tokenByBaseAcct: {
+              decimals: true,
+              symbol: true,
+              mint_acct: true,
+              name: true,
+              image_url: true,
+            },
+            tokenByQuoteAcct: {
+              decimals: true,
+              symbol: true,
+              mint_acct: true,
+              name: true,
+              image_url: true,
+            },
+            treasury_acct: true,
+            proposals_aggregate: {
+              aggregate: {
+                count: true,
+              },
+            },
+          },
+        },
+      });
+
+      const daoAggregates = dao_details.map<DaoAggregate>((daoDetails) =>
+        this.getDaoAggregateFromDaoDetailsQuery(daoDetails)
+      );
+
+      return daoAggregates;
+    } catch (e) {
+      console.error(e);
+      return this.rpcDaoClient.fetchAllDaos();
+    }
   }
-  async fetchDao(daoAddress: string): Promise<Dao | undefined> {
-    return undefined;
+
+  async fetchDao(daoSlug: string): Promise<DaoAggregate | null> {
+    try {
+      const { dao_details } = await this.graphqlClient.query({
+        dao_details: {
+          __args: {
+            where: {
+              slug: { _eq: daoSlug },
+            },
+            limit: 1,
+          },
+          name: true,
+          slug: true,
+          url: true,
+          github: true,
+          x_account: true,
+          description: true,
+          image_url: true,
+          daos: {
+            program_acct: true,
+            dao_acct: true,
+            tokenByBaseAcct: {
+              decimals: true,
+              symbol: true,
+              mint_acct: true,
+              name: true,
+              image_url: true,
+            },
+            tokenByQuoteAcct: {
+              decimals: true,
+              symbol: true,
+              mint_acct: true,
+              name: true,
+              image_url: true,
+            },
+            treasury_acct: true,
+            proposals_aggregate: {
+              aggregate: {
+                count: true,
+              },
+            },
+          },
+        },
+      });
+      if (dao_details[0]) {
+        return this.getDaoAggregateFromDaoDetailsQuery(dao_details[0]);
+      }
+      return null;
+    } catch (e) {
+      console.error(e);
+      // TODO: do we want to have as fallback here? We would need to pass in the protocol to this
+      // return this.rpcDaoClient.fetchDao(daoSlug, this.protocolMap.get(d.program_acct))
```

Comment:
I do like that.

Reply:
ah, I realize this is kind of difficult to implement. Because the RPC client doesn't actually allow fetching of a DAO by the slug, it can only fetch all the DAOs really. so we would have to fetch all DAOs again from RPC and then do a bunch of filtering. Do we maybe not want to do that?